portfolio/src/main/java/com/google/sps/servlets/UsernameManager.java

```diff
@@ -0,0 +1,97 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.gson.Gson;
+import com.google.sps.data.LoginStatus;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/login")
+public class UsernameManager extends HttpServlet {
+
+  private final String JSON_CONTENT_TYPE = "application/json";
+  private final String INDEX_PATH = "/";
+  // The object which connects to GCP datastore.
+  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+  // The object which converts Java to Json.
+  private final Gson gson = new Gson();
+  // The object which does user login/logout
+  private final UserService userService = UserServiceFactory.getUserService();
+  
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    response.setContentType(JSON_CONTENT_TYPE);
+    PrintWriter out = response.getWriter();
+    // If the user has logged in, send the pre-registered username to client,
+    // if not, send the login url.
+    String helperInfo = userService.isUserLoggedIn() 
+                      ? getUsername(userService.getCurrentUser().getUserId())
+                      : userService.createLoginURL(INDEX_PATH);
+    // Always send the logout url to client.
+    String logoutUrl = userService.createLogoutURL(INDEX_PATH);
+    out.println(gson.toJson(new LoginStatus(userService.isUserLoggedIn(), helperInfo, logoutUrl)));
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect(INDEX_PATH);
+      return;
+    }
+
+    String username = request.getParameter("username");
```

Comment:
Make a constant here and below.

Reply:
Fixed.
Sorry I forgot about that when I copied code from examples.